File: android_plugins/share/src/main/java/com/newvision/share/wrapper/ShareWrapper.java

/*
 * Project Name:ThirdPlugins
 * File Name:ShareWrapper.java
 * Package Name:
 * Date:17-1-9 下午2:33
 * Copyright (c) 2016, newvision All Rights Reserved.
 */

package com.newvision.share.wrapper;

import android.util.Log;

import com.alibaba.fastjson.JSON;
import com.newvision.share.Constants;
import com.unity3d.player.UnityPlayer;

/**
 * Created by gaozhenyu on 2016/12/20.
 * Unity调用部分，连接Unity中的C#
 */

public class ShareWrapper {
    private static ShareBase shareBase;

    /**
     * 初始化各个平台分享所需要的申请的AppID
     *
     * @param appids
     */
    public static void InitSharePlatform(String[] appids) {
        Constants.WECHATAPP_ID = appids[0];
        Constants.QQAPP_ID = appids[1];
        Constants.WEIBOAPP_ID = appids[2];
    }

    /**
     * 回调方法的自定义部分
     *
     * @param gameObjectName 接收回调脚本挂载的物体
     * @param methodName     接受回调的方法名
     */
    public static void CallBackMethod(String gameObjectName, String methodName) {
        Constants.GAMEOBJECT_NAME = gameObjectName;
        Constants.METHOD_NAME = methodName;
    }

    /**
     * 检测微信是否安装
     *
     * @return true->isInstalled false->not
     */
    public static boolean isWXAppInstalled() {
        return ShareManager.isWXInstalled();
    }

    /**
     * 微信登入，此方法会拉起微信=并放回code，或者取消授权
     */
    public static void LoginWithWechat() {
        LoginWithWechat.Login();
    }

    /**
     * 分享对外的调用方法
     *
     * @param contents 所要分享的内容
     * @param type     分享的类型 0，news 1、image
     * @param platform 分享平台 0、wechat friend 1、wechatmoment 2、QQ 3、weibo
     */
    public static void ShareContent(String contents, int type, int platform) {
        LoginWithWechat.isReciveCode = false;
        ShareContent content = JSON.parseObject(contents, ShareContent.class);
        switch (platform) {
            case 0:
                shareBase = new ShareToWechatFried();
                break;
            case 1:
                shareBase = new ShareToWechatMoment();
                break;
            case 2:
                shareBase = new ShareToQQ();
                break;
            case 3:
                shareBase = new ShareToWeibo();
                break;
        }
        shareBase.contents = content;
        switch (type) {
            case 0:
                shareBase.ShareNews();
                break;
            case 1:
                shareBase.ShareImage();
                break;
        }
    }

    public static void SendMessage(ErrorMsg msg) {
        String message = JSON.toJSONString(msg);
        Log.i("12312312321321", Constants.GAMEOBJECT_NAME + "-----" + Constants.METHOD_NAME + msg.errorCode);
        UnityPlayer.UnitySendMessage(Constants.GAMEOBJECT_NAME, Constants.METHOD_NAME, message);
    }

}

File: android_plugins/share/ShareWrapper.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class ShareWrapper : MonoBehaviour
{
    private static AndroidJavaClass shareWrapper;
    /// <summary>
    /// 初始化分享平台的key
    /// </summary>
    /// <param name="keys">0、微信 1、QQ 2、微博</param>
    public static void InitPlatformKey(string[] keys)
    {
        shareWrapper = new AndroidJavaClass("com.newvison.share.wrapper.ShareWrapper");
        shareWrapper.CallStatic("InitSharePlatform", new object[] { keys });
    }

    /// <summary>
    /// 分享内容，分享前需要调用 InitPlatformKey
    /// </summary>
    /// <param name="content">分享内容</param>
    /// <param name="type">分享的类型 0，news 1、image</param>
    /// <param name="platform">分享平台 0、wechat friend 1、wechatmoment 2、QQ 3、weibo</param>
    ///        ShareContent content = new ShareContent();
    ///        content.path = "path";
    ///        content.title = "123";
    ///        content.description = "123";
    ///        content.url = "http://www.baidu.com";
    ///        ShareWrapper.ShareTo(content, 0, 0);
    public static void ShareTo(ShareContent content, int type, int platform)
    {
        if (shareWrapper == null)
            Debug.LogError("please init share first");
        string contents = JsonUtility.ToJson(content);
        shareWrapper.CallStatic("ShareContent", new object[] { contents, type, platform });
    }
    public static void ShareTo(ShareContent content)
    {
        if (shareWrapper == null)
            Debug.LogError("please init share first");
        string contents = JsonUtility.ToJson(content);
        shareWrapper.CallStatic("ShareContent", new object[] { contents, (int)content.shareType, (int)content.plat });
    }
    /// <summary>
    /// 用微信登入
    /// </summary>
    public static void LoginWithWechat()
    {
        if (shareWrapper == null)
            Debug.LogError("please init share first");
        shareWrapper.CallStatic("LoginWithWechat");
    }
    /// <summary>
    /// 检测微信客户端是否安装
    /// </summary>
    /// <returns></returns>
    public static bool isWXInstalled()
    {
        if (shareWrapper == null)
            Debug.LogError("please init share first");
        return shareWrapper.CallStatic<bool>("isWXAppInstalled");
    }

    /// <summary>
    /// 回调方法参数
    /// </summary>
    /// <param name="gameObjectName">脚本挂载的物体，默认为“Main Camera”</param>
    /// <param name="methodName">回调方法，默认“ThirdPluginsCallBack”</param>
    public static void CallBackParms(string gameObjectName, string methodName)
    {
        shareWrapper.CallStatic("CallBackMethod", new object[] { gameObjectName, methodName });
    }

    /// <summary>
    /// 回调方法，可以根据 CallBackParms 放在其他对应的位置
    /// </summary>
    /// <param name="callbackJson"></param>
    public void ThirdPluginsCallBack(string callbackJson)
    {
        ErrorCode callbackCode = JsonUtility.FromJson<ErrorCode>(callbackJson);
        switch (callbackCode.errorCode)
        {
            case 0:
                if (callbackCode.loginCode != null)
                {

                }
                else
                {

                }
                break;
            case -1:
                break;
            case -2:
                break;
            case -3:
                break;
        }
    }

}
public class ShareContent
{
    public string path;
    public string title;
    public string description;
    public string url;
    public ShareType shareType;
    public SharePlatform plat;
}
public class ErrorCode
{
    public int errorCode;
    public string msg;
    public string loginCode;
}
public enum ShareType
{
    News = 0,
    Image = 1
}
public enum SharePlatform
{
    WechatFried = 0,
    WechatMoment = 1,
    QQ = 2,
    Weibo = 3
}